cEP-0019.md

```diff
@@ -0,0 +1,455 @@
+              "author": {
+                "login": "li-boxuan",
+                "url": "https://github.com/li-boxuan"
+              },
+              "reactions": {
+                "totalCount": 0,
+                "nodes": []
+              }
+            }
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+When a new issue/pull request is detected, gh-board will fetch that issue and
+loads it asynchronously. Meta-reviews also need to be fetched during this
+procedure.
+
+### Analysis of Meta-reviews
+
+This section explains how meta-review information can be utilized to
+generate a meaningful analysis.
+
+#### Ranking list
+
+A ranking list would be displayed on coala community website, according
+to points people get. The formula for points would be as follows:
+
+P1 = total number of THUMBS_UP 👍 a person gets for all reviews he did.
+
+P2 = total number of THUMBS_UP 👍 a person gives to other people for
+their reviews.
+
+N1 = total number of THUMBS_DOWN 👎 a person gets for all reviews he did.
+
+N2 = total number of THUMBS_DOWN 👎 a person gives to other people for
+their reviews.
+
+Note P2 and N2 are used to encourage people to get involved in this
+reactions feedback system.
+
+Then final score S would be:
+
+S = k * (P1 - N1) + c1 * P2 + c2 \* N2
+
+c1 = 0.2, c2 = 0.4, are coefficients. c2 is larger than c1 because people
+are usually reluctant to give negative reactions.
+
+k is the impact factor. The higher score a person has, the more impacts he has,
+thus his meta-reviews are more valuable. k is based on the scores in the
+previous iteration and is normalized. For example, in a previous iteration,
+Alice got 2 marks, Bob got 0.8 marks and Charlie got 10 marks. The calculation
+procedure would be as follows:
+
+```python
+>>> c = [2, 0.8, 10]
+>>> max_score = max(c)
+>>> min_score = min(c)
+>>> result = [(i - min_score) / (max_score - min_score) for i in c]
+>>> print(result)
+[0.13043478260869565, 0.0, 1.0]
+>>> result_adjust = [i + 0.1 for i in result]  # add bias
+>>> print(result_adjust)
+[0.23043478260869565, 0.1, 1.1]
+>>> result_rounded = [round(i, 3) for i in result_adjust]
+>>> print(result_rounded)
+[0.23, 0.1, 1.1]
+```
+
+Therefore, the k factor for Alice, Bob and Charlie in the next iteration would
+be: 0.23, 0.1 and 1.1, respectively.
+
+If a person has zeros for all factors, then he is regarded as an inactive
+reviewer and would not be displayed. If a person has negative score, he will
+still be displayed, but the meta-reviews he gives to other people will not
+count.
+
+Note the formular is a prototype and may be refined in practice.
+
+#### Meta-reviews in Need
+
+##### Definition
+
+When does a review need meta-review? There are two perspectives:
+
+1. Meta-review by developers, i.e. all people within the organization. The
```

Comment:
> Meta-review by developers, i.e. all people within the organization

That is confusing wrt coala terminology of newcomers.

Especially considering that this cEP proposes that newcomers will be required to do one meta-review and received a positive meta-review.

While you try to clarify it in the next sentence, it shows there is a fundamental problem with using these terms ;-)

Instead, these are meta-review during development of the PR, and meta-review after the merge.  In the future we might also include meta-review after the release.

Then note that meta-review after the merge will be limited to maintainers, and meta-reviews by others will not be counted.  This means that development meta-reviews are supposed to be done while development is happening, and not after the fact.  This helps the 'context' problem.

Also we need to factor in deleted comments here.  Need to include some background on who GitHub lets delete comments and when - I'm not familiar with those rules as I am always able to delete most comments, but some I can not.  However I know that when I am on other repos, I do get to delete comments, but I've never played with that.

We should allow people to delete their own review comments if it is done soon after the comment, but it is implicitly a bad comment if the author deletes it.  We could have automated meta-reviewing capture that.

Also editing a review comment after it has been seen by others (e.g. by email) is relevant for this project.

Where deleting/editing is definitely within scope of this project is *after* the review comment has been meta-reviewed.  Storing information in the community repo means that we can have history which is lost on Github.  We dont want to discard meta-review data simply because it was deleted from GitHub.  And if a comment has been meta-reviewed, and the author can delete their comment, they are destroying feedback from the community when they delete their comment.  That needs to warned against, and prevented, and possibly have a negative score given to it.

Also worth mentioning, but perhaps declaring out of scope, is self-meta-reviewing.  While the gh-board queue wouldnt promote this, it is easy to do in the GitHub UI, and can be quite effective to do that instead of a person deleting/editing their own comment.

Reply:
> We should allow people to delete their own review comments if it is done soon after the comment, but it is implicitly a bad comment if the author deletes it. We could have automated meta-reviewing capture that.

This doesn't sound realistic to me. We fetch reactions probably via a daily cron job, which means we are not able to monitor pull requests.